libkbfs/chat_rpc.go

```diff
@@ -0,0 +1,296 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"context"
+	"time"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/pkg/errors"
+)
+
+// CharRPC is an RPC based implementation for chat.
+type ChatRPC struct {
+	config   Config
+	log      logger.Logger
+	deferLog logger.Logger
+	client   chat1.LocalInterface
+}
+
+var _ rpc.ConnectionHandler = (*ChatRPC)(nil)
+
+// NewChatRPC constructs a new RPC based chat implementation.
+func NewChatRPC(config Config, kbCtx Context) *ChatRPC {
+	log := config.MakeLogger("")
+	deferLog := log.CloneWithAddedDepth(1)
+	c := &ChatRPC{
+		log:      log,
+		deferLog: deferLog,
+		config:   config,
+	}
+	conn := NewSharedKeybaseConnection(kbCtx, config, c)
+	c.client = chat1.LocalClient{Cli: conn.GetClient()}
+	return c
+}
+
+// HandlerName implements the ConnectionHandler interface.
+func (ChatRPC) HandlerName() string {
+	return "Chat"
+}
+
+// OnConnect implements the ConnectionHandler interface.
+func (c *ChatRPC) OnConnect(ctx context.Context, conn *rpc.Connection,
+	_ rpc.GenericClient, server *rpc.Server) error {
+	c.config.KBFSOps().PushConnectionStatusChange(KeybaseServiceName, nil)
+	return nil
+}
+
+// OnConnectError implements the ConnectionHandler interface.
+func (c *ChatRPC) OnConnectError(err error, wait time.Duration) {
+	c.log.Warning("Chat: connection error: %q; retrying in %s",
+		err, wait)
+	c.config.KBFSOps().PushConnectionStatusChange(KeybaseServiceName, err)
+}
+
+// OnDoCommandError implements the ConnectionHandler interface.
+func (c *ChatRPC) OnDoCommandError(err error, wait time.Duration) {
+	c.log.Warning("Chat: docommand error: %q; retrying in %s",
+		err, wait)
+	c.config.KBFSOps().PushConnectionStatusChange(KeybaseServiceName, err)
+}
+
+// OnDisconnected implements the ConnectionHandler interface.
+func (c *ChatRPC) OnDisconnected(_ context.Context,
+	status rpc.DisconnectStatus) {
+	if status == rpc.StartingNonFirstConnection {
+		c.log.Warning("Chat is disconnected")
+		c.config.KBFSOps().PushConnectionStatusChange(
+			KeybaseServiceName, errDisconnected{})
+	}
+}
+
+// ShouldRetry implements the ConnectionHandler interface.
+func (c *ChatRPC) ShouldRetry(_ string, _ error) bool {
+	return false
+}
+
+// ShouldRetryOnConnect implements the ConnectionHandler interface.
+func (c *ChatRPC) ShouldRetryOnConnect(err error) bool {
+	_, inputCanceled := err.(libkb.InputCanceledError)
+	return !inputCanceled
+}
+
+var _ Chat = (*ChatRPC)(nil)
+
+// Chat notes (will remove/rework these once the implementation is complete):
```

Comment:
No need to review these notes for correctness/grammar, I'll fix them in a later PR.  For now they're basically just notes to myself.